### ISSUE TITLE
Type physical quantities in editoast

### ISSUE THREAD

**Issue by leovalais** [OpenRailAssociation/osrd#9431]
Tracking issue: #9435 

# Goals

- Have a more reliable way than comments to handle units.
- Explicit unit changes
- Type check quantity mismatch (e.g. `speed = length * time` shouldn't compile)
- Get rid of all manual conversions (millimeters to meters, etc)

# Constraints

- Should work with the current railjson format (which can change of course, but one migration at a time)
- Can work for integer and floating point arithmetic 

# Proposed solution: [`uom`](https://crates.io/crates/uom)

Benefits:

- Wide adoption
- Old, but still maintained
- Convenient API (a bit opinionated on the naming conventions)
- Supports all numeric types for quantities (even rational numbers!)
- Out of the box `serde` support

I experimented a bit with the `serde` interface of uom. We can't configure what is the unit the data we're (de)serializing: e.g. when serializing an `uom::si::u64::Length`, `uom` always serializes the value in `meters`. This is a bit annoying for us since we have weird units in our railjson format (`mm`, `km/s`, `minutes/100km` 🤯). Thankfully `serde` has a [feature](https://serde.rs/remote-derive.html) that will greatly limit the amount of boilerplate needed in most cases.

```rust
#[derive(Debug, Deserialize, Serialize)]
#[serde(remote = "uom::si::f64::Length")]
struct Millimeters(#[serde(getter = "uom::si::f64::Length::get::<millimeter>")] f64);

impl From<Millimeters> for uom::si::f64::Length {
    fn from(mm: Millimeters) -> Self {
        uom::si::f64::Length::new::<millimeter>(mm.0)
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub struct UndirectedTrackRange {
    pub track_section: String,
    #[serde(with = "Millimeters")]
    pub begin: uom::si::f64::Length,
    #[serde(with = "Millimeters")]
    pub end: uom::si::f64::Length,
}
```

The approach of `uom` is great for our case since we only have to deal with units when we're importing, exporting or visualizing data. But when we use the quantity for computations, we only care about what the quantity is and not which unit we used to input it.

# Units in the RailJson format

> [!NOTE]
> This part is not in scope for this issue, but rather to initiate a discussion on how to manage quantities in the railjson format in general.

We use different units to represent the same quantity across our schemas (eg. we both use `mm` and `m` for distances, `s` and `min` for time, etc.). That might be nice to manually edit railjson files, but the inconsistency isn't great, especially since it leaks beyond the parsing step.

If we were to bring some consistency to the format, I believe we'd first need to determine whether "RailJson" is *an exchange format between OSRD processes*, or a *user-facing format that allows __convenient manual editing__*.

If we consider it a purely technical format, that can be readable but not necessarily tailored for manual edition, I'd suggest to turn *all quantities into their base unit* (lengths in meters, durations in seconds, masses in grams, efforts in newtons, etc.). In places where it makes sense, we can avoid using floats without losing accuracy using rational numbers, which `uom` supports. We'd loose *a little* in readability but improve consistency and reduce complexity by a great deal.

On the other hand, if railjson should be editable manually, we still have to revise the ergonomics of quantities. As of now, in order to edit a quantity, we have to either know its unit or look it up. It's not great. And units change. The numeric representation does too. To have something more convenient and robust to change, I think we'd have to accept multiple value formats to streamline the edition ; like, for lengths: `12` (implicitly meters), `0.5` (still meters), `"50 cm"`, `1/2 km`, `"1/3 l.y."` (one third of a lightyear), etc. In this case, we obviously need a parsing step which isn't provided by `uom`, but further justify the need to abstract away from units.

As fun as expressing track section lengths in lightyears sounds, I'm partial towards using the base SI unit for quantities...

------

Before we start working on implementing this feature (with `uom` or not), I think it's important to also figure out how we want to handle units for our inputs, and if anyone has different opinions and/or priorities on this topic.

**Comment by leovalais** [comment]
### Recap of a small workshop on the subject with @Tristramg @woshilapin @leovalais 

* Let's use `uom`
* Traction / towed rolling stocks seem like good candidates to start with as there's a bunch of different units involved and they're not as entangled as infra schemas.
* The RailJson schema must not change.
* Editoast inputs / outputs must not change (front, core, import and osm interfaces)

Interrogation: should we roll our own unit framework to avoid dealing with `uom` serialization issues? Apparently not, the adapters using `serde(remote = ` seem easy enough to write to not justify a custom framework for that issue alone. To be confirmed when the first `uom` PRs land.
